modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java

```diff
@@ -110,23 +115,27 @@ private void requestStart(ChannelHandlerContext ctx) {
             // this looks like a malformed request and will forward without validation
             ctx.channel().eventLoop().submit(() -> forwardFullRequest(ctx));
         } else {
-            validator.apply(httpRequest, ctx.channel(), new ActionListener<>() {
-                @Override
-                public void onResponse(Void unused) {
-                    // Always use "Submit" to prevent reentrancy concerns if we are still on event loop
-                    ctx.channel().eventLoop().submit(() -> forwardFullRequest(ctx));
-                }
-
-                @Override
-                public void onFailure(Exception e) {
-                    // Always use "Submit" to prevent reentrancy concerns if we are still on event loop
-                    ctx.channel().eventLoop().submit(() -> forwardRequestWithDecoderExceptionAndNoContent(ctx, e));
-                }
-            });
+            Transports.assertDefaultThreadContext(threadContext);
+            // this prevents thread-context changes to propagate to the validation listener
+            // atm, the validation listener submits to the event loop executor, which doesn't know about the ES thread-context,
+            // so this is just a defensive play, in case the code inside the listener changes to not use the event loop executor
+            ContextPreservingActionListener<Void> contextPreservingActionListener = new ContextPreservingActionListener<>(
+                threadContext.wrapRestorable(threadContext.newStoredContext()),
+                ActionListener.wrap(aVoid ->
+                // Always use "Submit" to prevent reentrancy concerns if we are still on event loop
+                ctx.channel().eventLoop().submit(() -> forwardFullRequest(ctx)),
+                    e -> ctx.channel().eventLoop().submit(() -> forwardRequestWithDecoderExceptionAndNoContent(ctx, e))
+                )
+            );
+            // this prevents thread-context changes to propagate beyond the validation, as netty worker threads are reused
+            try (ThreadContext.StoredContext ignore = threadContext.newStoredContext()) {
+                validator.apply(httpRequest, ctx.channel(), contextPreservingActionListener);
+            }
         }
     }
 
     private void forwardFullRequest(ChannelHandlerContext ctx) {
+        Transports.assertDefaultThreadContext(threadContext);
```

Comment:
++ on the asserts here (and anywhere for the thread context - it is very helpful). however, could be outside the context of this PR but it would be great if only had 1 definition of what default thread context means. 

org.elasticsearch.transport.Transports#assertDefaultThreadContext defines it as 
```java
    assert threadContext.getRequestHeadersOnly().isEmpty()
            || REQUEST_HEADERS_ALLOWED_ON_DEFAULT_THREAD_CONTEXT.containsAll(threadContext.getRequestHeadersOnly().keySet())
            : "expected empty context but was " + threadContext.getRequestHeadersOnly() + " on " + Thread.currentThread().getName();
```
and org.elasticsearch.common.util.concurrent.ThreadContext#isDefaultContext defines it as 
```java
return threadLocal.get() == DEFAULT_CONTEXT;
```            
I understand they are slightly different, but ideally they would not be and we would only have a single definition.  Not needed to address on this PR , just wanted to get your thoughts if it is possible to collapse these 2 into 1.

Reply:
> ideally they would not be and we would only have a single definition.

++ Looked into this briefly too. I would've expected `threadLocal.get() == DEFAULT_CONTEXT` to be used everywhere inplace of `threadContext.getRequestHeadersOnly().isEmpty() ....`. I tried it and it failed intermitently on one of the license IT, where the thread context contained the license expiration warning **response header**. I think that response headers might be the reason for the different ways to check the default context....
I haven't investigated further... it might be a bug with response headers leaking....

I'll make a note to follow-up on this when I'll feel out the woods...